src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs

```diff
@@ -39,10 +42,49 @@ public static IMvcCoreBuilder AddMvcCore(this IServiceCollection services)
                 throw new ArgumentNullException(nameof(services));
             }
 
+            var partManager = GetApplicationPartManager(services);
+            services.TryAddSingleton(partManager);
+
             ConfigureDefaultServices(services);
             AddMvcCoreServices(services);
 
-            return new MvcCoreBuilder(services);
+            var builder = new MvcCoreBuilder(services, partManager);
+
+            return builder;
+        }
+
+        private static ApplicationPartManager GetApplicationPartManager(IServiceCollection services)
+        {
+            var manager = GetServiceFromCollection<ApplicationPartManager>(services);
+            if (manager == null)
+            {
+                var environment = GetServiceFromCollection<IHostingEnvironment>(services);
+                if (environment == null)
+                {
+                    // Consider just returning an empty application part collection. Makes
+                    // tests easier and its not a condition we expect in production scenarios
+                    // where IHostingEnvironment will always be part of the ServiceCollection
+                    // we get passed in.
+                    throw new InvalidOperationException();
+                }
+
+                var assemblies = new DefaultAssemblyProvider(environment).CandidateAssemblies;
+                manager = new ApplicationPartManager();
+
+                foreach (var assembly in assemblies)
+                {
+                    manager.ApplicationParts.Add(new AssemblyPart(assembly));
+                }
+            }
+
+            return manager;
+        }
+
+        private static T GetServiceFromCollection<T>(IServiceCollection services)
```

Comment:
We might need to come up with some alternative here - this code worries me because it's diving into implementation details of a completely unrelated component.